# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2023-05-14T05:08:46.268155905+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyi764xqw/model/ with prefix bf808d865d214b4d
[INFO 2023-05-14T05:08:46.278404638+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:08:46.278645806+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2023-05-14T05:08:50.824555819+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1dfpvd6b/model/ with prefix f77faab3c47e4baf
[INFO 2023-05-14T05:08:50.834229156+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2023-05-14T05:08:50.834307255+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmprezvnqws as temporary training directory


[INFO 2023-05-14T05:10:57.222907568+00:00 kernel.cc:1214] Loading model from path /tmp/tmprezvnqws/model/ with prefix 1b047aa69d7a49c8
[INFO 2023-05-14T05:10:57.237684015+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2023-05-14T05:10:57.237884537+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2023-05-14T05:10:57.237929332+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2023-05-14T05:11:15.583069174+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp1ja0dla/model/ with prefix 0412be34ba964185
[INFO 2023-05-14T05:11:15.587750626+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:11:15.587794057+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2023-05-14T05:11:17.09189488+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_oqoqsda/model/ with prefix 3a42597e801b4a58
[INFO 2023-05-14T05:11:17.112314136+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2023-05-14T05:11:18.190786191+00:00 kernel.cc:1214] Loading model from path /tmp/tmpztrqp0j9/model/ with prefix ebeae8bb1da44c5d
[INFO 2023-05-14T05:11:18.195440702+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2023-05-14T05:11:19.799228616+00:00 kernel.cc:1214] Loading model from path /tmp/tmptfz5du93/model/ with prefix d438255faf1a4ab0
[INFO 2023-05-14T05:11:19.826824468+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2023-05-14T05:11:20.899531207+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoiztkg9v/model/ with prefix 7f4431c85df047cc
[INFO 2023-05-14T05:11:20.905558802+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2023-05-14T05:11:21.906000099+00:00 kernel.cc:1214] Loading model from path /tmp/tmp61n_layc/model/ with prefix c759b1819cfd4942
[INFO 2023-05-14T05:11:21.909314194+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2023-05-14T05:11:23.11149841+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj5j65q6w/model/ with prefix 3ef6a41a5f3f44eb
[INFO 2023-05-14T05:11:23.119578716+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2023-05-14T05:11:24.610267102+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6d1i856u/model/ with prefix 70e9a88eb4974075
[INFO 2023-05-14T05:11:24.63131826+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2023-05-14T05:11:25.850063005+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9dg7umux/model/ with prefix e6afd69022d84a09
[INFO 2023-05-14T05:11:25.860443576+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:11:25.860505389+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2023-05-14T05:11:27.65780092+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcat9dy6w/model/ with prefix 18f20b37bd7d4ab8
[INFO 2023-05-14T05:11:27.674800261+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2023-05-14T05:11:28.77237484+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk0xjoqgi/model/ with prefix 67691782571a42c3
[INFO 2023-05-14T05:11:28.778432723+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2023-05-14T05:11:30.109259893+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj69_pap3/model/ with prefix 3195df0a84f549fd
[INFO 2023-05-14T05:11:30.124776135+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2023-05-14T05:11:31.202776593+00:00 kernel.cc:1214] Loading model from path /tmp/tmp88a4p5kq/model/ with prefix 953d83aa64624519
[INFO 2023-05-14T05:11:31.209061909+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2023-05-14T05:11:32.416827206+00:00 kernel.cc:1214] Loading model from path /tmp/tmp66dl5_vr/model/ with prefix ec90206779f84d90
[INFO 2023-05-14T05:11:32.428566863+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2023-05-14T05:11:33.998576123+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1hla6ubn/model/ with prefix 3cfa41921066418e
[INFO 2023-05-14T05:11:34.005099273+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2023-05-14T05:11:35.117399538+00:00 kernel.cc:1214] Loading model from path /tmp/tmpafpmp1d9/model/ with prefix 01af23f141e344dc
[INFO 2023-05-14T05:11:35.125421438+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2023-05-14T05:11:36.397507782+00:00 kernel.cc:1214] Loading model from path /tmp/tmplct07k9b/model/ with prefix 652237f3e4154ff4
[INFO 2023-05-14T05:11:36.410742334+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:11:36.410795131+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2023-05-14T05:11:37.681585866+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwqhivxha/model/ with prefix 6c97b4c8a8f44b1b
[INFO 2023-05-14T05:11:37.695316492+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2023-05-14T05:11:39.378845547+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwop2fop0/model/ with prefix c48c378adadd4176
[INFO 2023-05-14T05:11:39.390899691+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2023-05-14T05:11:40.850694508+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg0lms28n/model/ with prefix cb90a2f4fd384afd
[INFO 2023-05-14T05:11:40.869492889+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2023-05-14T05:11:42.223542481+00:00 kernel.cc:1214] Loading model from path /tmp/tmpi0iinv4d/model/ with prefix b4f437493963403a
[INFO 2023-05-14T05:11:42.239903135+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2023-05-14T05:11:43.738608397+00:00 kernel.cc:1214] Loading model from path /tmp/tmppe8jr3k7/model/ with prefix 0bcfddfb012a45bd
[INFO 2023-05-14T05:11:43.744344514+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2023-05-14T05:11:44.818739083+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmfb4wj4c/model/ with prefix b0af7f208ca5438c
[INFO 2023-05-14T05:11:44.824928487+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2023-05-14T05:11:46.006025234+00:00 kernel.cc:1214] Loading model from path /tmp/tmpaqrnij_2/model/ with prefix d80d7fd6cb114520
[INFO 2023-05-14T05:11:46.015152595+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2023-05-14T05:11:47.649399657+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq9yku7be/model/ with prefix fba7b0c3048f4346
[INFO 2023-05-14T05:11:47.655342336+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:11:47.655382874+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2023-05-14T05:11:48.872236922+00:00 kernel.cc:1214] Loading model from path /tmp/tmpur85kauh/model/ with prefix c8916c00299647bc
[INFO 2023-05-14T05:11:48.884544762+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2023-05-14T05:11:50.090804221+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6noizjtu/model/ with prefix b3140c9fc8ff4df8
[INFO 2023-05-14T05:11:50.101161696+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2023-05-14T05:11:51.223396734+00:00 kernel.cc:1214] Loading model from path /tmp/tmpiflm_0go/model/ with prefix 4afde3a4adb04e8e
[INFO 2023-05-14T05:11:51.230276195+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2023-05-14T05:11:52.881164982+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoenlcs0t/model/ with prefix 2bd064b8f0694fcd
[INFO 2023-05-14T05:11:52.886409005+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2023-05-14T05:11:54.205674752+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6tce8_bd/model/ with prefix 9ba5159258c44390
[INFO 2023-05-14T05:11:54.218978125+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2023-05-14T05:11:55.898628743+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc8igcr83/model/ with prefix 122a55ff7f35465e
[INFO 2023-05-14T05:11:55.92709145+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2023-05-14T05:11:57.717359738+00:00 kernel.cc:1214] Loading model from path /tmp/tmpokp481kl/model/ with prefix bd56fbeeba92458f
[INFO 2023-05-14T05:11:57.729537313+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:11:57.729585383+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2023-05-14T05:11:59.399639102+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9ijctk3i/model/ with prefix 27888cd588f842e1
[INFO 2023-05-14T05:11:59.406530217+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2023-05-14T05:12:00.723047377+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwwl5tl84/model/ with prefix fbe80a26ee874825
[INFO 2023-05-14T05:12:00.737163573+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2023-05-14T05:12:01.937108021+00:00 kernel.cc:1214] Loading model from path /tmp/tmp36ia8qba/model/ with prefix 59973e85e2214480
[INFO 2023-05-14T05:12:01.950035701+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2023-05-14T05:12:03.131598842+00:00 kernel.cc:1214] Loading model from path /tmp/tmphqv7_0ws/model/ with prefix a2607550d03745be
[INFO 2023-05-14T05:12:03.140030529+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2023-05-14T05:12:05.047193621+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8ss1ci0b/model/ with prefix 4cd66f3afc30499c
[INFO 2023-05-14T05:12:05.062702045+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2023-05-14T05:12:06.259180713+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_lyw3icr/model/ with prefix 055b29af2ca446d3
[INFO 2023-05-14T05:12:06.268584273+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2023-05-14T05:12:08.095406105+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7lvaooi4/model/ with prefix 47c7ea6ceb63423a
[INFO 2023-05-14T05:12:08.109398222+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:12:08.109444523+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2023-05-14T05:12:09.828678026+00:00 kernel.cc:1214] Loading model from path /tmp/tmpex2pvc7v/model/ with prefix 0c856de185dd4f30
[INFO 2023-05-14T05:12:09.842272283+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2023-05-14T05:12:10.885631171+00:00 kernel.cc:1214] Loading model from path /tmp/tmphljqk_rp/model/ with prefix 8d3ae400ef27454a
[INFO 2023-05-14T05:12:10.890382214+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2023-05-14T05:12:12.221017006+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv6emuvam/model/ with prefix 04399b90b54346a0
[INFO 2023-05-14T05:12:12.237159755+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2023-05-14T05:12:13.457247633+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxve9t98_/model/ with prefix 9ff42fafe6544d7b
[INFO 2023-05-14T05:12:13.467257398+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2023-05-14T05:12:15.510296397+00:00 kernel.cc:1214] Loading model from path /tmp/tmpasxgcjbi/model/ with prefix 5d8cf8651004406b
[INFO 2023-05-14T05:12:15.528936692+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2023-05-14T05:12:16.938193005+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuyok82fo/model/ with prefix e4f9510b6ef944cf
[INFO 2023-05-14T05:12:16.947986231+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2023-05-14T05:12:17.966266616+00:00 kernel.cc:1214] Loading model from path /tmp/tmpclqgzkzf/model/ with prefix 2972cbafb05b400a
[INFO 2023-05-14T05:12:17.969959695+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2023-05-14T05:12:19.29987719+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjhu_62cq/model/ with prefix 9b63deb374744d7a
[INFO 2023-05-14T05:12:19.314746417+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:12:19.314792541+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2023-05-14T05:12:21.001458935+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzk6hjsqw/model/ with prefix 4193a1d6252b46d8
[INFO 2023-05-14T05:12:21.015247623+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2023-05-14T05:12:22.522197564+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdx7ei535/model/ with prefix 0e9f2c7ded604514
[INFO 2023-05-14T05:12:22.526931438+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2023-05-14T05:12:23.637935412+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoudltn5u/model/ with prefix f27b7fee09954b8a
[INFO 2023-05-14T05:12:23.645296327+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2023-05-14T05:12:24.945325888+00:00 kernel.cc:1214] Loading model from path /tmp/tmp93gyhou_/model/ with prefix e76d011e3cdf4ec7
[INFO 2023-05-14T05:12:24.958623033+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2023-05-14T05:12:26.33638635+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_fkoz6c9/model/ with prefix 2f128603e66c4dd6
[INFO 2023-05-14T05:12:26.35246678+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2023-05-14T05:12:27.856586421+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk5g7x22t/model/ with prefix 8dce0e0f47c04883
[INFO 2023-05-14T05:12:27.864681103+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2023-05-14T05:12:28.994026914+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe4zsziim/model/ with prefix 7796e4001b684c18
[INFO 2023-05-14T05:12:29.002267345+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2023-05-14T05:12:30.053539184+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0u9oqzs3/model/ with prefix c563720c6ee0417f
[INFO 2023-05-14T05:12:30.056986324+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:12:30.057026131+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2023-05-14T05:12:31.399355564+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_teq3nuu/model/ with prefix 223e70a92c584f80
[INFO 2023-05-14T05:12:31.41453904+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2023-05-14T05:12:33.308333175+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn69k162s/model/ with prefix ba8f14343d3941cc
[INFO 2023-05-14T05:12:33.32261341+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2023-05-14T05:12:34.46599472+00:00 kernel.cc:1214] Loading model from path /tmp/tmpiigicz96/model/ with prefix 49b6b369a4a2407f
[INFO 2023-05-14T05:12:34.470787148+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2023-05-14T05:12:35.667010044+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2qee0cs0/model/ with prefix d937300228cc42cf
[INFO 2023-05-14T05:12:35.674300949+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2023-05-14T05:12:36.902910865+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7abrp5a1/model/ with prefix 5c03e67fbd954b9e
[INFO 2023-05-14T05:12:36.913719537+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2023-05-14T05:12:38.156155548+00:00 kernel.cc:1214] Loading model from path /tmp/tmpss_c7dez/model/ with prefix 7ba15922d9e64dfb
[INFO 2023-05-14T05:12:38.166741511+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2023-05-14T05:12:39.261894031+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8_i43ix3/model/ with prefix f14c936d083247e9
[INFO 2023-05-14T05:12:39.267407401+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2023-05-14T05:12:40.928228585+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0qhksfdl/model/ with prefix 4fa2acb0992f46c7
[INFO 2023-05-14T05:12:40.955993553+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:12:40.95604739+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2023-05-14T05:12:42.161742142+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4xvc26ob/model/ with prefix 6439aa1105284b0d
[INFO 2023-05-14T05:12:42.171628205+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2023-05-14T05:12:43.323548443+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe980wk72/model/ with prefix 8ad9c03a863f4ad2
[INFO 2023-05-14T05:12:43.331468532+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2023-05-14T05:12:44.995709144+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbhvaqqww/model/ with prefix d384711647394d57
[INFO 2023-05-14T05:12:45.000661136+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2023-05-14T05:12:46.171770706+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_b1e1576/model/ with prefix 9877ef248fc44367
[INFO 2023-05-14T05:12:46.178167835+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2023-05-14T05:12:48.190705379+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjv7ybqgc/model/ with prefix 8d6866adffaa4338
[INFO 2023-05-14T05:12:48.210866025+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2023-05-14T05:12:49.504430616+00:00 kernel.cc:1214] Loading model from path /tmp/tmp61i4uw56/model/ with prefix 92be60200faa4366
[INFO 2023-05-14T05:12:49.51680304+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2023-05-14T05:12:50.659811947+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0dii_xs7/model/ with prefix 7822be1f2d044311
[INFO 2023-05-14T05:12:50.666700594+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2023-05-14T05:12:51.849453188+00:00 kernel.cc:1214] Loading model from path /tmp/tmpku0tqbpe/model/ with prefix 0fc5e63d4dc140a3
[INFO 2023-05-14T05:12:51.858360716+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:12:51.858407241+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2023-05-14T05:12:53.023344919+00:00 kernel.cc:1214] Loading model from path /tmp/tmppm79sckg/model/ with prefix 9778f4f9fc5a46b4
[INFO 2023-05-14T05:12:53.030966934+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2023-05-14T05:12:55.006157482+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl90fu0ng/model/ with prefix 34503b919ee44858
[INFO 2023-05-14T05:12:55.022357272+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2023-05-14T05:12:56.541574503+00:00 kernel.cc:1214] Loading model from path /tmp/tmpits4x0jt/model/ with prefix 9be9cb4010124d69
[INFO 2023-05-14T05:12:56.54889111+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2023-05-14T05:12:58.424475108+00:00 kernel.cc:1214] Loading model from path /tmp/tmp43r8dbmv/model/ with prefix 317d9fd5296d41b9
[INFO 2023-05-14T05:12:58.436627751+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2023-05-14T05:13:00.105794557+00:00 kernel.cc:1214] Loading model from path /tmp/tmpaq3r7xh2/model/ with prefix b1c7e78e755a4b57
[INFO 2023-05-14T05:13:00.114195099+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2023-05-14T05:13:01.173690146+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq8qngpt2/model/ with prefix 96f7f9214daa4228
[INFO 2023-05-14T05:13:01.177923829+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2023-05-14T05:13:02.259774922+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfeikb38w/model/ with prefix 5f284c1ad8154864
[INFO 2023-05-14T05:13:02.264731288+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:13:02.264787217+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2023-05-14T05:13:03.413178956+00:00 kernel.cc:1214] Loading model from path /tmp/tmp402t37ib/model/ with prefix 3be44c9eeb8d4c90
[INFO 2023-05-14T05:13:03.420792577+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2023-05-14T05:13:05.201643007+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkw06mriw/model/ with prefix b7d89a420664446f
[INFO 2023-05-14T05:13:05.209537583+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2023-05-14T05:13:07.03298542+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf9fygtws/model/ with prefix fa9ebe76380d44e0
[INFO 2023-05-14T05:13:07.043451223+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2023-05-14T05:13:08.872113775+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw5duu34x/model/ with prefix c4599c4f98fc4d1a
[INFO 2023-05-14T05:13:08.884490102+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2023-05-14T05:13:10.113173445+00:00 kernel.cc:1214] Loading model from path /tmp/tmpikdyebu3/model/ with prefix 19f3c9d955564c33
[INFO 2023-05-14T05:13:10.123839204+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2023-05-14T05:13:11.323888581+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2l0qz_88/model/ with prefix a47537ae80ad4d88
[INFO 2023-05-14T05:13:11.333109651+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2023-05-14T05:13:13.518244953+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo3wgxkqr/model/ with prefix c048fb72376f478c
[INFO 2023-05-14T05:13:13.53811638+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:13:13.538178061+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2023-05-14T05:13:14.737236535+00:00 kernel.cc:1214] Loading model from path /tmp/tmpas58dyot/model/ with prefix 323f5a897c3d4560
[INFO 2023-05-14T05:13:14.744836334+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2023-05-14T05:13:16.180942484+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq10b707j/model/ with prefix af1365d4be174e90
[INFO 2023-05-14T05:13:16.198589887+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2023-05-14T05:13:18.051093882+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuo5e7fo9/model/ with prefix 4ecfd2723994443c
[INFO 2023-05-14T05:13:18.071257571+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2023-05-14T05:13:19.412630664+00:00 kernel.cc:1214] Loading model from path /tmp/tmphw_jtyoh/model/ with prefix 11b979856dc64754
[INFO 2023-05-14T05:13:19.425206735+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2023-05-14T05:13:20.523004638+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd7pju8s0/model/ with prefix 520bfb0e19e841eb
[INFO 2023-05-14T05:13:20.528538984+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2023-05-14T05:13:21.755126176+00:00 kernel.cc:1214] Loading model from path /tmp/tmprs03ms9u/model/ with prefix 94e786049db34563
[INFO 2023-05-14T05:13:21.766377852+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2023-05-14T05:13:22.923496088+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj5_2hjv4/model/ with prefix 19a8650a77864380
[INFO 2023-05-14T05:13:22.930978609+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2023-05-14T05:13:24.363477324+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy3rwjvxa/model/ with prefix 2331efd70f3e45dc
[INFO 2023-05-14T05:13:24.382215351+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-05-14T05:13:24.382264235+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2023-05-14T05:13:25.641508521+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwp5a4ka6/model/ with prefix bf905dc682c84683
[INFO 2023-05-14T05:13:25.652270386+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2023-05-14T05:13:27.287029688+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_ev76ooj/model/ with prefix a53fe99a293149d7
[INFO 2023-05-14T05:13:27.294436112+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2023-05-14T05:13:28.482649339+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3v1ozi_1/model/ with prefix d671719459a34913
[INFO 2023-05-14T05:13:28.491525347+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2023-05-14T05:13:29.715108814+00:00 kernel.cc:1214] Loading model from path /tmp/tmpigioozsv/model/ with prefix 45cf60f0e0574861
[INFO 2023-05-14T05:13:29.72562431+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2023-05-14T05:13:30.866216681+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2q06_sck/model/ with prefix 6ab4f178ca04481a
[INFO 2023-05-14T05:13:30.872024462+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2023-05-14T05:13:32.038425281+00:00 kernel.cc:1214] Loading model from path /tmp/tmpiidkrq6w/model/ with prefix 775d8dc2ea414b1a
[INFO 2023-05-14T05:13:32.046578296+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2023-05-14T05:13:33.402075939+00:00 kernel.cc:1214] Loading model from path /tmp/tmptylmz11a/model/ with prefix dc98595815e9424b
[INFO 2023-05-14T05:13:33.416664411+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv
